### DISCUSSION TITLE
Tranformation, and BDFACs Diva experiment files support

### DISCUSSION THREAD

**Discussion by jainaditi0204** [whitews/FlowKit#102] [Q&A] [answered]
Hi Scott, 

I have been working with FlowKit and gatingMLs, as you know. As per your suggestion, I added transformations to the gate dimensions, and exported a gatingML file. However, on reading it back, it doesn't show any transformations linked to the existing gating strategy. Also, in the gatingML file, the only information associated with the added transforms is its type (name), and the parameter values are missing. Is it how it is represented in the files exported via FlowKit? 

The second point is that, are there any future plans of importing BDFACs Diva experiment files in FlowKit directly? It might eliminate the dependency on third party softwares to extract gating information if it could simply be retrieved from the original file. Also, the use case could be generalised, in a way, that users will be able to retrieve the information of their interest from the experiment files attained via FACs Diva. I am curious to hear your thoughts on it!

Again, thank you very much.

Best,
Aditi

**Comment by whitews** [comment]
Hi Aditi,

What type of transform are you adding? GatingML only supports certain ones. 

There are no current plans to support FACs Diva experiment files. Supporting additional non-standardized (i.e. open & documented) application formats is not on our current roadmap. The best chance any XML-based format would have in being supported would be to provide an XML Schema Definition (XSD) file.

-Scott

**Comment by jainaditi0204** [comment]
Hi Scott, 

I am adding linear and logicle transforms to the gate dimensions. I have defined them separately with parameters, and then I add them in string form, as you suggested. 
Here's an example:

linear_transform = fk.transforms.LinearTransform('flin', param_t=1, param_a=0)
dimension_k = fk.Dimension('FSC-H', range_min=17830.373046875, range_max=176089.5625, transformation_ref='linear_transform')

I'm hoping this is the correct way of adding them? I also tried using 'flin' as the transformation_ref, but it still doesn't show anything when I import the gatingML. 

Best,
Aditi

**Comment by whitews** [comment]
Okay, the linear and logicle transforms are compatible with GatingML and should work. Are you sure you are also adding the transform itself to the Session?

Say you are working in a Session using a sample group named "my_group" (you can omit the sample group stuff if you are using the 'default' sample group). Then you would add your linear transform example like so:

```
sample_group = "my_group"

# define a transform and add it to the session
linear_transform = fk.transforms.LinearTransform('flin', param_t=1, param_a=0)
session.add_transform(linear_transform, group_name=sample_group)
```

From there you'd create dimensions with the transform reference (a.k.a. the ID 'flin') to use for gates. Then add the gates to the session.

The best example of doing all this is in this notebook:

https://github.com/whitews/FlowKit/blob/master/examples/flowkit-session-replicate-flowjo-wsp.ipynb

The first part of that notebook just imports a workspace file to show as an example. The rest of the notebook replicates the entire gating hierarchy from scratch.

**Comment by jainaditi0204** [comment]
Ah, okay. I added the transforms to the session, and they show up now :)

Thank you for clearing it up! 
And thank you again for FlowKit. It's really helpful :)

Best,
Aditi

**Comment by whitews** [comment, marked as answer]
Glad it's working now. I'll see if I can add a final validation check when exporting to check for missing references.
